File: agent-web/src/api/permission-groups.ts
/**
 * Permission Group API 模块：对接后端 `/permission-groups` 前缀的权限组资源
 * 全部端点（`app/api/v1/permission_groups.py`，5 个端点）。
 *
 * 权限组是命名的工具审批集合，AgentApp 可绑定权限组以复用审批配置。
 * 两层权限系统：
 * 1. permission_preset（枚举）：none / strict / destructive_only
 * 2. permission_group_id（外键）：指向本模块管理的命名权限组
 *
 * 解析优先级（assembly.py resolve_interrupt_on）：
 * permission_group_id > permission_preset > interrupt_on（legacy）
 *
 * 约定：
 * - 响应信封 {code, message, data} 已由 request.ts 拦截器解包，本模块函数
 *   返回值即 data 载荷；
 * - 全量列表端点（GET /permission-groups）返回裸数组；分页端点
 *   （GET /permission-groups/page）返回 PageResult<PermissionGroupRow>；
 * - 路径参数 group_id 为整数主键，无需 encodeURIComponent；
 * - name 全局唯一，创建重名后端 422。
 */
import { del, get, patch, post } from '@/utils/request'
import type { PageQuery, PageResult } from '@/types'

/** 权限预设枚举值（与后端 PERMISSION_PRESETS 一致） */
export type PermissionPreset = 'none' | 'strict' | 'destructive_only'

/** 权限组资产行（对应后端 PermissionGroupRead 全字段） */
export interface PermissionGroupRow {
  id: number
  name: string
  /** 需要人工审批的工具名称列表 */
  tool_names: string[]
  description: string
  created_by: string | null
}

/**
 * 权限组创建 payload（POST /permission-groups）。
 * name 必填且全局唯一；tool_names 默认为空列表；description 默认为空字符串。
 */
export interface PermissionGroupCreatePayload {
  name: string
  tool_names?: string[]
  description?: string
}

/**
 * 权限组部分更新 payload（PATCH /permission-groups/{group_id}，name 不可改）。
 * 空 payload 后端 422 拒绝（"nothing to update"）。
 */
export interface PermissionGroupPatchPayload {
  tool_names?: string[]
  description?: string
}

/** 把 PageQuery 透传为后端查询参数（page/pageSize/keyword） */
function toParams(query: PageQuery): Record<string, unknown> {
  return { page: query.page, pageSize: query.pageSize, keyword: query.keyword }
}

/** 全量列表：GET /permission-groups（按 id 升序的裸数组） */
export function listPermissionGroups(): Promise<PermissionGroupRow[]> {
  return get<PermissionGroupRow[]>('/permission-groups')
}

/** 分页列表：GET /permission-groups/page（keyword 模糊匹配 name/description） */
export function listPermissionGroupsPage(
  query: PageQuery = {},
): Promise<PageResult<PermissionGroupRow>> {
  return get<PageResult<PermissionGroupRow>>('/permission-groups/page', { params: toParams(query) })
}

/** 单条详情：GET /permission-groups/{group_id}（不存在 404） */
export function getPermissionGroup(groupId: number): Promise<PermissionGroupRow> {
  return get<PermissionGroupRow>(`/permission-groups/${groupId}`)
}

/** 创建权限组：POST /permission-groups（201；重名 422） */
export function createPermissionGroup(
  payload: PermissionGroupCreatePayload,
): Promise<PermissionGroupRow> {
  return post<PermissionGroupRow>('/permission-groups', payload)
}

/** 部分更新：PATCH /permission-groups/{group_id}（name 不可改） */
export function patchPermissionGroup(
  groupId: number,
  payload: PermissionGroupPatchPayload,
): Promise<PermissionGroupRow> {
  return patch<PermissionGroupRow>(`/permission-groups/${groupId}`, payload)
}

/** 删除权限组：DELETE /permission-groups/{group_id} */
export function deletePermissionGroup(groupId: number): Promise<null> {
  return del<null>(`/permission-groups/${groupId}`)
}

File: app/api/v1/permission_groups.py
"""Admin API for permission group assets (CRUD).

Phase-1 scope: all assets are globally shared (no per-user ownership checks);
every endpoint only authenticates via ``get_current_user`` and records the
creator in the audit-only ``created_by`` field.

Error semantics: missing resources return 404; name collisions return 422;
unexpected failures return 500 after ``logger.exception``.
"""

from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query, Request
from sqlmodel import Session as DBSession
from sqlmodel import col, or_, select

from app.api.v1.agent_assets_common import (
    _read_patch_body,
    _validate_payload,
    get_db_session,
)
from app.api.v1.auth import get_current_user
from app.core.config import settings
from app.core.limiter import limiter
from app.core.logging import logger
from app.models.agent_assets import PermissionGroup
from app.models.user import User
from app.schemas.agent_apps import (
    PermissionGroupCreate,
    PermissionGroupRead,
    PermissionGroupUpdate,
)
from app.schemas.base import ApiResponse, PageResult

router = APIRouter()


def _get_group_or_404(db: DBSession, group_id: int) -> PermissionGroup:
    """Fetch one permission group row or raise 404."""
    group = db.get(PermissionGroup, group_id)
    if group is None:
        raise HTTPException(status_code=404, detail=f"permission group '{group_id}' not found")
    return group


@router.get("/permission-groups", response_model=ApiResponse[list[PermissionGroupRead]])
@limiter.limit(settings.RATE_LIMIT_ENDPOINTS["agent_app"][0])
async def list_permission_groups(
    request: Request,
    db: DBSession = Depends(get_db_session),
    user: User = Depends(get_current_user),
) -> ApiResponse[Any]:
    """List every stored permission group.

    Args:
        request: The FastAPI request object for rate limiting.
        db: Request-scoped DB session.
        user: Authenticated user resolved from the user access token.

    Returns:
        Envelope carrying all permission group rows ordered by name.
    """
    try:
        return ApiResponse.success(list(db.exec(select(PermissionGroup).order_by(col(PermissionGroup.name))).all()))
    except Exception as exc:
        logger.exception("permission_group_list_failed")
        raise HTTPException(status_code=500, detail=str(exc)) from exc


@router.post("/permission-groups", response_model=ApiResponse[PermissionGroupRead], status_code=201)
@limiter.limit(settings.RATE_LIMIT_ENDPOINTS["agent_app"][0])
async def create_permission_group(
    request: Request,
    payload: PermissionGroupCreate,
    db: DBSession = Depends(get_db_session),
    user: User = Depends(get_current_user),
) -> ApiResponse[Any]:
    """Create a named permission group.

    Args:
        request: The FastAPI request object for rate limiting.
        payload: The permission group definition.
        db: Request-scoped DB session.
        user: Authenticated user used for audit attribution.

    Returns:
        Envelope carrying the persisted permission group row.

    Raises:
        HTTPException: 422 when the name is already taken.
    """
    try:
        existing = db.exec(
            select(PermissionGroup).where(col(PermissionGroup.name) == payload.name)
        ).first()
        if existing is not None:
            logger.warning("permission_group_create_conflict", name=payload.name)
            raise HTTPException(status_code=422, detail=f"permission group '{payload.name}' already exists")

        group = PermissionGroup(
            name=payload.name,
            tool_names=payload.tool_names,
            description=payload.description,
            created_by=user.username or str(user.id),
        )
        db.add(group)
        db.commit()
        db.refresh(group)
        logger.info("permission_group_created", group_id=group.id, name=payload.name)
        return ApiResponse.success(group, code=201)
    except HTTPException:
        raise
    except Exception as exc:
        logger.exception("permission_group_create_failed", name=payload.name)
        raise HTTPException(status_code=500, detail=str(exc)) from exc


@router.get("/permission-groups/page", response_model=ApiResponse[PageResult[PermissionGroupRead]])
@limiter.limit(settings.RATE_LIMIT_ENDPOINTS["agent_app"][0])
async def list_permission_groups_page(
    request: Request,
    page: int = Query(1, ge=1),
    page_size: int = Query(10, ge=1, le=100, alias="pageSize"),
    keyword: str | None = Query(None, max_length=200),
    db: DBSession = Depends(get_db_session),
    user: User = Depends(get_current_user),
) -> ApiResponse[Any]:
    """List permission groups with server-side pagination.

    Args:
        request: The FastAPI request object for rate limiting.
        page: 1-based page number.
        page_size: Rows per page (exposed as ``pageSize``).
        keyword: Optional case-insensitive substring matched against name or description.
        db: Request-scoped DB session.
        user: Authenticated user resolved from the user access token.

    Returns:
        Envelope carrying a PageResult of permission group rows ordered by name.
    """
    try:
        stmt = select(PermissionGroup)
        count_stmt = select(PermissionGroup)
        if keyword:
            like = f"%{keyword}%"
            predicate = or_(
                col(PermissionGroup.name).ilike(like),
                col(PermissionGroup.description).ilike(like),
            )
            stmt = stmt.where(predicate)
            count_stmt = count_stmt.where(predicate)

        total = len(db.exec(count_stmt).all())
        rows = db.exec(stmt.order_by(col(PermissionGroup.name)).offset((page - 1) * page_size).limit(page_size)).all()

        return ApiResponse.success(
            PageResult(items=list(rows), total=total, page=page, page_size=page_size)
        )
    except Exception as exc:
        logger.exception("permission_group_list_page_failed")
        raise HTTPException(status_code=500, detail=str(exc)) from exc


@router.get("/permission-groups/{group_id}", response_model=ApiResponse[PermissionGroupRead])
@limiter.limit(settings.RATE_LIMIT_ENDPOINTS["agent_app"][0])
async def get_permission_group(
    request: Request,
    group_id: int,
    db: DBSession = Depends(get_db_session),
    user: User = Depends(get_current_user),
) -> ApiResponse[Any]:
    """Fetch one permission group by id.

    Args:
        request: The FastAPI request object for rate limiting.
        group_id: Permission group primary key.
        db: Request-scoped DB session.
        user: Authenticated user resolved from the user access token.

    Returns:
        Envelope carrying the matching permission group row.

    Raises:
        HTTPException: 404 when the permission group does not exist.
    """
    try:
        return ApiResponse.success(_get_group_or_404(db, group_id))
    except HTTPException:
        raise
    except Exception as exc:
        logger.exception("permission_group_read_failed", group_id=group_id)
        raise HTTPException(status_code=500, detail=str(exc)) from exc


@router.patch("/permission-groups/{group_id}", response_model=ApiResponse[PermissionGroupRead])
@limiter.limit(settings.RATE_LIMIT_ENDPOINTS["agent_app"][0])
async def update_permission_group(
    request: Request,
    group_id: int,
    db: DBSession = Depends(get_db_session),
    user: User = Depends(get_current_user),
) -> ApiResponse[Any]:
    """Partially update a permission group (name is immutable).

    Args:
        request: The FastAPI request object for rate limiting.
        group_id: Permission group primary key.
        db: Request-scoped DB session.
        user: Authenticated user resolved from the user access token.

    Returns:
        Envelope carrying the updated permission group row.

    Raises:
        HTTPException: 404 when missing, 422 on empty payload.
    """
    body = await _read_patch_body(request)
    payload = _validate_payload(PermissionGroupUpdate, body)
    try:
        group = db.get(PermissionGroup, group_id)
        if group is None:
            raise HTTPException(status_code=404, detail=f"permission group '{group_id}' not found")

        updates = payload.model_dump(exclude_unset=True)
        if not updates:
            raise HTTPException(status_code=422, detail="nothing to update")

        for field, value in updates.items():
            setattr(group, field, value)

        db.add(group)
        db.commit()
        db.refresh(group)
        logger.info("permission_group_updated", group_id=group_id)
        return ApiResponse.success(group)
    except HTTPException:
        raise
    except Exception as exc:
        logger.exception("permission_group_update_failed", group_id=group_id)
        raise HTTPException(status_code=500, detail=str(exc)) from exc


@router.delete("/permission-groups/{group_id}", response_model=ApiResponse[None])
@limiter.limit(settings.RATE_LIMIT_ENDPOINTS["agent_app"][0])
async def delete_permission_group(
    request: Request,
    group_id: int,
    db: DBSession = Depends(get_db_session),
    user: User = Depends(get_current_user),
) -> ApiResponse[None]:
    """Delete a permission group.

    Args:
        request: The FastAPI request object for rate limiting.
        group_id: Permission group primary key.
        db: Request-scoped DB session.
        user: Authenticated user resolved from the user access token.

    Returns:
        Envelope with null data on successful deletion.

    Raises:
        HTTPException: 404 when the permission group does not exist.
    """
    try:
        group = _get_group_or_404(db, group_id)
        db.delete(group)
        db.commit()
        logger.info("permission_group_deleted", group_id=group_id)
        return ApiResponse.success(None)
    except HTTPException:
        raise
    except Exception as exc:
        logger.exception("permission_group_delete_failed", group_id=group_id)
        raise HTTPException(status_code=500, detail=str(exc)) from exc
